File: backend/internal/factor/dsl/ast.go
// Package dsl implements the factor DSL engine (Go).
// Grammar EBNF: see docs/spec/factor-dsl.md.
package dsl

// Node is an AST node.
type Node interface {
	node()
}

// NumberLit: 42, 3.14, -1.5e-3
type NumberLit struct{ Value float64 }

// BoolLit: true, false
type BoolLit struct{ Value bool }

// StringLit: "EURUSD"
type StringLit struct{ Value string }

// FieldRef: $open, $close, $volume
type FieldRef struct{ Name string }

// UnaryExpr: -x, !x
type UnaryExpr struct {
	Op   string // "-" or "!"
	Expr Node
}

// BinaryExpr: a + b, a > b, a && b
type BinaryExpr struct {
	Op          string
	Left, Right Node
}

// TernaryExpr: cond ? a : b
type TernaryExpr struct {
	Cond, True, False Node
}

// CallExpr: func(args...)
type CallExpr struct {
	Name string
	Args []Node
}

// FactorRef: user-defined factor (no $ prefix)
type FactorRef struct{ Name string }

func (*NumberLit) node()   {}
func (*BoolLit) node()     {}
func (*StringLit) node()   {}
func (*FieldRef) node()    {}
func (*UnaryExpr) node()   {}
func (*BinaryExpr) node()  {}
func (*TernaryExpr) node() {}
func (*CallExpr) node()    {}
func (*FactorRef) node()   {}

File: research/ant_research/factor/dsl/ast_.py
"""ant factor DSL — AST node types.

Ported from alfq_research. Aligned with Go backend/go/internal/factor/dsl/ast.go.
"""

from dataclasses import dataclass, field


@dataclass
class NumberLit:
    value: float

@dataclass
class BoolLit:
    value: bool

@dataclass
class StringLit:
    value: str

@dataclass
class FieldRef:
    name: str

@dataclass
class FactorRef:
    name: str

@dataclass
class UnaryExpr:
    op: str
    expr: "Node"

@dataclass
class BinaryExpr:
    op: str
    left: "Node"
    right: "Node"

@dataclass
class TernaryExpr:
    cond: "Node"
    true_expr: "Node"
    false_expr: "Node"

@dataclass
class CallExpr:
    name: str
    args: list["Node"] = field(default_factory=list)


Node = NumberLit | BoolLit | StringLit | FieldRef | FactorRef | UnaryExpr | BinaryExpr | TernaryExpr | CallExpr
